Controller/ReceivedUrl.php

```diff
@@ -0,0 +1,250 @@
+                $this->_redirect($redirectUrl);
+            } catch (NoSuchEntityException $noSuchEntityException) {
+                $logMessage = $noSuchEntityException->getMessage();
+                $this->logHelper->addInfoLog('NoSuchEntityException: ' . $logMessage);
+
+                $this->bugsnag->registerCallback(function ($report) use ($incrementId) {
+                    $report->setMetaData([
+                        'order incrementId' => $incrementId,
+                    ]);
+                });
+                $this->bugsnag->notifyError('NoSuchEntityException: ', $logMessage);
+
+                $errorMessage = __('Something went wrong. Please contact the seller.');
+                $this->messageManager->addErrorMessage($errorMessage);
+
+                $this->_redirect($this->getErrorRedirectUrl());
+            } catch (LocalizedException $e) {
+                $logMessage = $e->getMessage();
+                $this->logHelper->addInfoLog('LocalizedException:' . $logMessage);
+
+                $errorMessage = __('Something went wrong. Please contact the seller.');
+                $this->messageManager->addErrorMessage($errorMessage);
+
+                $this->bugsnag->registerCallback(function ($report) use ($boltSignature, $boltPayload) {
+                    $report->setMetaData([
+                        'bolt_signature' => $boltSignature,
+                        'bolt_payload'   => $boltPayload,
+                    ]);
+                });
+                $this->bugsnag->notifyError('LocalizedException: ', $logMessage);
+                $this->_redirect($this->getErrorRedirectUrl());
+            }
+        } else {
+            // Potentially it is attack.
+            $logMessage = 'bolt_signature and Magento signature are not equal';
+            $this->logHelper->addInfoLog($logMessage);
+
+            $this->bugsnag->registerCallback(function ($report) use ($boltSignature, $boltPayload) {
+                $report->setMetaData([
+                    'bolt_signature' => $boltSignature,
+                    'bolt_payload'   => $boltPayload
+                ]);
+            });
+            $this->bugsnag->notifyError('OrderReceivedUrl Error', $logMessage);
+
+            $errorMessage = __('Something went wrong. Please contact the seller.');
+            $this->messageManager->addErrorMessage($errorMessage);
+            $this->_redirect($this->getErrorRedirectUrl());
+        }
+    }
+
+    /**
+     * @param $payload
+     * @return mixed
+     */
+    private function getReferenceFromPayload($payload)
+    {
+        $payloadArray = json_decode($payload, true);
+        return  $payloadArray['transaction_reference'];
+    }
+
+    /**
+     * @param $payload
+     * @return mixed
+     */
+    private function getIncrementIdFromPayload($payload)
+    {
+        $payloadArray = json_decode($payload, true);
+        $displayId = $payloadArray['display_id'];
+
+        list($incrementId, $quoteId) = array_pad(
```

Comment:
Why do we bother with a pad here to ensure an assigned `$quoteId` when we don't use it? This could just be `$incrementId = strtok($displayId, ' / ');`

Reply:
This happens multiple times through the PR